### PR TITLE
Added example notebook for horovod training

### PR DESCRIPTION
**Pull request by uditbhatia**
The structure is created and Tested, some formatting and documentation is left.

### CODE REVIEW

File: sagemaker-python-sdk/tensorflow_script_mode_horovod/mnist_hvd.py [hunk 1]
```diff
@@ -0,0 +1,125 @@
+from __future__ import print_function
```
[the remark this answers is not in the document]
Comment (uditbhatia) [reply sagemaker-python-sdk/tensorflow_script_mode_horovod/mnist_hvd.py]: Done

File: sagemaker-python-sdk/tensorflow_script_mode_horovod/mnist_hvd.py [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+
+if __name__ == '__main__':
+    
+    num_gpus = int(os.environ['SM_NUM_GPUS'])
+
+    parser = argparse.ArgumentParser()
+
+    # Data, model, and output directories. These are required.
+    parser.add_argument('--output-dir', type=str, default=os.environ['SM_OUTPUT_DIR'])
+    parser.add_argument('--model-dir', type=str, default=os.environ['SM_MODEL_DIR'])
```
[the remark this answers is not in the document]
Comment (uditbhatia) [reply sagemaker-python-sdk/tensorflow_script_mode_horovod/mnist_hvd.py]: Done

File: sagemaker-python-sdk/tensorflow_script_mode_horovod/mnist_hvd.py [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+    num_classes = 10
+
+    # Horovod: adjust number of epochs based on number of GPUs.
+    epochs = int(math.ceil(12.0 / hvd.size()))
+
+    # Input image dimensions
+    img_rows, img_cols = 28, 28
+
+    # The data, shuffled and split between train and test sets
+    # (x_train, y_train), (x_test, y_test) = mnist.load_data()
```
[the remark this answers is not in the document]
Comment (uditbhatia) [reply sagemaker-python-sdk/tensorflow_script_mode_horovod/mnist_hvd.py]: Done

File: sagemaker-python-sdk/tensorflow_script_mode_horovod/tensorflow_script_mode_horovod.ipynb [hunk 4]
```diff
@@ -0,0 +1,468 @@
...
+    "\n",
+    "import keras\n",
+    "from keras.datasets import mnist\n",
+    "(x_train, y_train), (x_test, y_test) = mnist.load_data()\n",
+    "\n",
+    "s3_train_path = \"s3://{}/mnist/train.npz\".format(default_s3_bucket)\n",
+    "s3_test_path = \"s3://{}/mnist/test.npz\".format(default_s3_bucket)\n",
+    "\n",
+    "try:\n",
+    "    # Create local directory\n",
```
[the remark this answers is not in the document]
Comment (uditbhatia) [reply sagemaker-python-sdk/tensorflow_script_mode_horovod/tensorflow_script_mode_horovod.ipynb]: Removed

File: sagemaker-python-sdk/tensorflow_script_mode_horovod/tensorflow_script_mode_horovod.ipynb [hunk 5]
```diff
@@ -0,0 +1,468 @@
...
+   "source": [
+    "Horovod is a distributed training framework based on MPI. You can find more details at [Horovod README](https://github.com/uber/horovod).\n",
+    "\n",
+    "Horovod Distributed Training can be perfomed on SageMaker using the SageMaker Tensorflow container. SageMaker creates the MPI environment and executes the `mpirun` command to execute the training script.\n",
+    "\n",
+    "MPI environment for Horovod can be configured by following flags in SageMaker SDK:\n",
+    "\n",
+    "* ``enabled (bool)``: If set to ``True``, the MPI setup is performed and ``mpirun`` command is executed.\n",
+    "* ``processes_per_host (int) [Optional]``: Number of processes MPI should launch on each host. Note, this should not be greater than the available slots on the selected instance type.\n",
+    "* ``custom_mpi_options (str) [Optional]``: Additional command line arguments that we need to pass to ``mpirun``.\n",
```
[the remark this answers is not in the document]
Comment (uditbhatia) [reply sagemaker-python-sdk/tensorflow_script_mode_horovod/tensorflow_script_mode_horovod.ipynb]: Done

File: sagemaker-python-sdk/tensorflow_script_mode_horovod/tensorflow_script_mode_horovod.ipynb [hunk 6]
```diff
@@ -0,0 +1,468 @@
...
+    "\n",
+    "In this example notebook, we will create a MNIST horovod training job."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Set up the environment"
```
[the remark this answers is not in the document]
Comment (uditbhatia) [reply sagemaker-python-sdk/tensorflow_script_mode_horovod/tensorflow_script_mode_horovod.ipynb]: Done

File: sagemaker-python-sdk/tensorflow_script_mode_horovod/tensorflow_script_mode_horovod.ipynb [hunk 7]
```diff
@@ -0,0 +1,468 @@
...
+    "sagemaker_iam_role = \"SageMakerRole\"\n",
+    "train_script = \"mnist_hvd.py\"\n",
+    "instance_count = 2"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Download the dataset and save to local `/tmp` directory and upload the dataset to s3"
```
[the remark this answers is not in the document]
Comment (uditbhatia) [reply sagemaker-python-sdk/tensorflow_script_mode_horovod/tensorflow_script_mode_horovod.ipynb]: done, I will consult @eslesar-aws too

File: sagemaker-python-sdk/tensorflow_script_mode_horovod/tensorflow_script_mode_horovod.ipynb [hunk 8]
```diff
@@ -0,0 +1,468 @@
...
+    "    print('test data at ', s3_test_path)\n",
+    "finally:\n",
+    "    shutil.rmtree('/tmp/data')\n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Construct a script for horovod distributed training"
```
[the remark this answers is not in the document]
Comment (uditbhatia) [reply sagemaker-python-sdk/tensorflow_script_mode_horovod/tensorflow_script_mode_horovod.ipynb]: done

File: sagemaker-python-sdk/tensorflow_script_mode_horovod/tensorflow_script_mode_horovod.ipynb [hunk 9]
```diff
@@ -0,0 +1,468 @@
...
+    "estimator = TensorFlow(entry_point=train_script,\n",
+    "                       role=sagemaker_iam_role,\n",
+    "                       train_instance_count=instance_count,\n",
+    "                       train_instance_type=train_instance_type,\n",
+    "                       script_mode=True,\n",
+    "                       framework_version='1.12',\n",
+    "                       distributions=distributions,\n",
+    "                       security_group_ids=['sg-0919a36a89a15222f'],\n",
+    "                       subnets=['subnet-0c07198f3eb022ede', 'subnet-055b2819caae2fd1f'],\n",
+    "                       base_job_name='hvd-mnist-vpc')"
```
[the remark this answers is not in the document]
Comment (uditbhatia) [reply sagemaker-python-sdk/tensorflow_script_mode_horovod/tensorflow_script_mode_horovod.ipynb]: Done

File: sagemaker-python-sdk/tensorflow_script_mode_horovod/tensorflow_script_mode_horovod.ipynb [hunk 10]
```diff
@@ -0,0 +1,468 @@
...
+   "outputs": [],
+   "source": [
+    "estimator.fit({\"train\":s3_train_path, \"test\":s3_test_path})"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Horovod Training in SageMaker using VPC environment\n",
```
[the remark this answers is not in the document]
Comment (uditbhatia) [reply sagemaker-python-sdk/tensorflow_script_mode_horovod/tensorflow_script_mode_horovod.ipynb]: Done

File: sagemaker-python-sdk/tensorflow_script_mode_horovod/vpc_infra_cfn.json [hunk 11]
```diff
@@ -0,0 +1,242 @@
+{
```
[the remark this answers is not in the document]
Comment (uditbhatia) [reply sagemaker-python-sdk/tensorflow_script_mode_horovod/vpc_infra_cfn.json]: Can you please share any example of copyright in json file ?

[the remark this answers is not in the document]
Comment (uditbhatia) [reply sagemaker-python-sdk/tensorflow_script_mode_horovod/vpc_infra_cfn.json]: Discussed offline and its not required.

File: sagemaker-python-sdk/tensorflow_script_mode_horovod/mnist_hvd.py [hunk 12]
```diff
@@ -0,0 +1,125 @@
...
+    model.fit(x_train, y_train,
+              batch_size=batch_size,
+              callbacks=callbacks,
+              epochs=epochs,
+              verbose=1,
+              validation_data=(x_test, y_test))
+    score = model.evaluate(x_test, y_test, verbose=0)
+
+    print('Test loss:', score[0])
+    print('Test accuracy:', score[1])
```
[the remark this answers is not in the document]
Comment (uditbhatia) [reply sagemaker-python-sdk/tensorflow_script_mode_horovod/mnist_hvd.py]: Done

File: sagemaker-python-sdk/tensorflow_script_mode_horovod/tensorflow_script_mode_horovod.ipynb [hunk 13]
```diff
@@ -0,0 +1,574 @@
...
+    "In this example we are using the [Keras MNIST horovod example](https://github.com/uber/horovod/blob/master/examples/keras_mnist.py) from the horovod github repository.\n",
+    "\n",
+    "In order to run this script we have to make following modifications:"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### 1. Accept `--model_dir` command-line argument\n",
```
Comment (mvsusp) [inline sagemaker-python-sdk/tensorflow_script_mode_horovod/tensorflow_script_mode_horovod.ipynb]: Do you mind detailing here that model_dir point to /opt/ml/model which is the special folder that SageMaker saves in the end of training and that works for serving as well. Please mentioned that it allows the training job to integrates with other SageMaker services and avoid the model to be lost after training given that the training job is ephemeral.

Please add some links as well.

File: sagemaker-python-sdk/tensorflow_script_mode_horovod/tensorflow_script_mode_horovod.ipynb [hunk 14]
```diff
@@ -0,0 +1,574 @@
...
+    "parser = argparse.ArgumentParser()\n",
+    "parser.add_argument('--model_dir', type=str)\n",
+    "```"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### 2. Load train and test data\n",
```
Comment (mvsusp) [inline sagemaker-python-sdk/tensorflow_script_mode_horovod/tensorflow_script_mode_horovod.ipynb]: Please link to the documentation mentioning these variables and explain how they are configured.

File: sagemaker-python-sdk/tensorflow_script_mode_horovod/tensorflow_script_mode_horovod.ipynb (hunk 13 above)
Comment (uditbhatia) [reply sagemaker-python-sdk/tensorflow_script_mode_horovod/tensorflow_script_mode_horovod.ipynb]: Done

File: sagemaker-python-sdk/tensorflow_script_mode_horovod/tensorflow_script_mode_horovod.ipynb (hunk 14 above)
Comment (uditbhatia) [reply sagemaker-python-sdk/tensorflow_script_mode_horovod/tensorflow_script_mode_horovod.ipynb]: Done

**Pull request merged**